app/services/region_report_service.rb

```diff
@@ -42,24 +43,40 @@ def compile_control_and_registration_data
   # the previous three quarters. Each quarter cohort is made up of patients registered
   # in the previous quarter who has had a follow up visit in the current quarter.
   def compile_cohort_trend_data
-    Quarter.new(date: selected_date).downto(3).each do |results_quarter|
-      cohort_quarter = results_quarter.previous_quarter
+    Rails.cache.fetch(cohort_cache_key, version: cohort_cache_version, expires_in: 7.days, force: force_cache?) do
+      Quarter.new(date: selected_date).downto(3).each do |results_quarter|
+        cohort_quarter = results_quarter.previous_quarter
 
-      period = {cohort_period: :quarter,
-                registration_quarter: cohort_quarter.number,
-                registration_year: cohort_quarter.year}
-      query = MyFacilities::BloodPressureControlQuery.new(facilities: @facilities, cohort_period: period)
-      @data[:quarterly_registrations] << {
-        results_in: format_quarter(results_quarter),
-        patients_registered: format_quarter(cohort_quarter),
-        registered: query.cohort_registrations.count,
-        controlled: query.cohort_controlled_bps.count,
-        no_bp: query.cohort_missed_visits_count,
-        uncontrolled: query.cohort_uncontrolled_bps.count
-      }.with_indifferent_access
+        period = {cohort_period: :quarter,
+                  registration_quarter: cohort_quarter.number,
+                  registration_year: cohort_quarter.year}
+        query = MyFacilities::BloodPressureControlQuery.new(facilities: @facilities, cohort_period: period)
+        @data[:quarterly_registrations] << {
+          results_in: format_quarter(results_quarter),
+          patients_registered: format_quarter(cohort_quarter),
+          registered: query.cohort_registrations.count,
+          controlled: query.cohort_controlled_bps.count,
+          no_bp: query.cohort_missed_visits_count,
+          uncontrolled: query.cohort_uncontrolled_bps.count
+        }.with_indifferent_access
+      end
     end
   end
 
+  private
+
+  def cohort_cache_key
+    "#{self.class}/cohort_trend_data/#{region.model_name}/#{region.id}/#{organizations.map(&:id)}/#{selected_date.to_s(:iso8601)}"
```

Comment:
Nitpick: isn't organization higher in the hierarchy than region? This seems backwards.

Reply:
Yeah, was just following the order set in the other keys. Its not meant to be strictly hierarchical, just easy to read and adjust if need be.